apps/state/lib/state/schedule.ex

```diff
@@ -171,9 +171,23 @@ defmodule State.Schedule do
   defp build_filter_matchers(%{stops: stops, trips: trips} = filters) do
```

Comment:
What about making a matcher for `%{stops: stops, routes: routes}`? Then you wouldn't need to back out from trip to route?

Reply:
I've tried to do it but it works slower.. 

```bash
Levs-MBP:~ lev$ for i in {1..10}; do curl -g "http://localhost:4000/schedules?flter[date]=2019-06-18&filter[max_time]=10:44&filter[min_time]=7:44&filter[route]=22,44,23,34,45,14,36,15,38,28,19,41,42,10,32&filter[stop]=15661,15660,1583,1341,1582,1340,390,1581,1580,392,393,1181,394,395,1180,1338,396,1579,1337,1578,1335,1577,399,1334,1576,1498,1575,1179,1332,1497,1574,1331,1573,15820,13321,1339,1351,1350,1346,1345,401,1586,1344,1585,568,1584,1342,404,405,21151,406,407,40001,410,411,412,1330,1572,1571,1570,1174,1173,1172,1569,386,1567,387,1566,388,1485,1328&include=route,trip,stop,prediction,vehicle&api_key=1341e103b7fa4fb58dae4248c6899acf" -s -o /dev/null -w  "%{time_starttransfer}\n"; done
2.219786
2.191719
2.253387
2.123809
2.135842
2.072874
2.101699
2.264730
2.266189
2.193600
```

This is how I did it: 

```elixir
  defp build_filter_matchers(%{stops: stops, routes: routes} = filters) do
    stop_sequence_matchers = build_stop_sequence_matchers(filters[:stop_sequence])

    matchers =
      for stop_id <- stops,
          route_id <- routes,
          stop_sequence_matcher <- stop_sequence_matchers do
        stop_sequence_matcher
        |> Map.put(:route_id, route_id)
        |> Map.put(:stop_id, stop_id)
      end

    %{index: :stop_id, matchers: matchers}
  end
```
Do you have ideas how to implement it better?

Reply:
That's the right idea, but also including the part where you use `StopsOnRoute` to ensure that we only build a matcher when the stop is actually on the route.

Reply:
It's still slow (no performance improvements comparing to unchanged version) like that: 

```bash
Levs-MBP:~ lev$ for i in {1..10}; do curl -g "http://localhost:4000/schedules?flter[date]=2019-06-18&filter[max_time]=10:44&filter[min_time]=7:44&filter[route]=22,44,23,34,45,14,36,15,38,28,19,41,42,10,32&filter[stop]=15661,15660,1583,1341,1582,1340,390,1581,1580,392,393,1181,394,395,1180,1338,396,1579,1337,1578,1335,1577,399,1334,1576,1498,1575,1179,1332,1497,1574,1331,1573,15820,13321,1339,1351,1350,1346,1345,401,1586,1344,1585,568,1584,1342,404,405,21151,406,407,40001,410,411,412,1330,1572,1571,1570,1174,1173,1172,1569,386,1567,387,1566,388,1485,1328&include=route,trip,stop,prediction,vehicle&api_key=1341e103b7fa4fb58dae4248c6899acf" -s -o /dev/null -w  "%{time_starttransfer}\n"; done
1.510134
1.385606
1.576934
1.653379
1.528449
1.371150
1.360028
1.348777
1.358736
1.346985
```

with 

```elixir
  defp build_filter_matchers(%{stops: stops, routes: routes} = filters) do
    stop_sequence_matchers = build_stop_sequence_matchers(filters[:stop_sequence])

    filtered_routes =
      stops
      |> State.RoutesAtStop.by_stops_and_direction()
      |> MapSet.new()
      |> MapSet.intersection(MapSet.new(routes))

    matchers =
      for stop_id <- stops,
          route_id <- filtered_routes,
          stop_sequence_matcher <- stop_sequence_matchers do
        stop_sequence_matcher
        |> Map.put(:route_id, route_id)
        |> Map.put(:stop_id, stop_id)
      end

    %{index: :stop_id, matchers: matchers}
  end
```

Reply:
Can you put `RoutesAtStop` into the loop, something like:
```elixir
  defp build_filter_matchers(%{stops: stops, routes: routes} = filters) do
    stop_sequence_matchers = build_stop_sequence_matchers(filters[:stop_sequence])
    route_set = MapSet.new(routes)

    matchers =
      for stop_id <- stops,
          routes_at_stop = MapSet.new(State.RoutesAtStop.by_stops_and_direction([stop_id]),
          route_id <- MapSet.intersection(routes_at_stop, route_set),
          stop_sequence_matcher <- stop_sequence_matchers do
        stop_sequence_matcher
        |> Map.put(:route_id, route_id)
        |> Map.put(:stop_id, stop_id)
      end

    %{index: :stop_id, matchers: matchers}

Reply:
> Then you wouldn't need to back out from trip to route?

I think we cannot do it as filters always include date ([ref](https://github.com/mbta/api/blob/master/apps/api_web/lib/api_web/controllers/schedule_controller.ex#L128)). It means we still need to figure out the trips for this particular date and then filter them by the routes.